File: androidApp/src/main/java/com/tora/yetanotherpomo/ui/components/HoldToConfirmButton.kt
package com.tora.yetanotherpomo.ui.components

import androidx.compose.animation.core.Animatable
import androidx.compose.animation.core.tween
import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.gestures.detectTapGestures
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.hapticfeedback.HapticFeedbackType
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.platform.LocalHapticFeedback
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import com.tora.yetanotherpomo.ui.theme.CaprasimoFamily
import com.tora.yetanotherpomo.ui.theme.Organic
import kotlinx.coroutines.launch

/**
 * The doc's "Hold to end" control: an outlined pill whose fill sweeps left-to-right over
 * [holdDurationMs] while pressed, firing [onHoldComplete] at 100%; releasing early cancels and
 * eases the fill back to zero. Ports `startHold`/`holdUp` from the reference prototype 1:1.
 */
@Composable
fun HoldToConfirmButton(
    text: String,
    holdDurationMs: Int,
    onHoldComplete: () -> Unit,
    modifier: Modifier = Modifier,
    minHeight: Dp = 56.dp,
    borderColor: Color = Organic.colors.lockedOnSurface.copy(alpha = Organic.opacity.border),
    fillColor: Color = Organic.colors.accentRamp.c800,
    contentColor: Color = Organic.colors.lockedOnSurface.copy(alpha = Organic.opacity.strong),
) {
    val progress = remember { Animatable(0f) }
    val scope = rememberCoroutineScope()
    val haptics = LocalHapticFeedback.current
    val shape = RoundedCornerShape(percent = 50)
    val motion = Organic.motion
    val borderWidth = Organic.stroke.hairline

    Box(
        modifier = modifier
            .fillMaxWidth()
            .height(minHeight)
            .clip(shape)
            .border(BorderStroke(borderWidth, borderColor), shape)
            .pointerInput(holdDurationMs) {
                detectTapGestures(
                    onPress = {
                        val holdJob = scope.launch {
                            progress.snapTo(0f)
                            progress.animateTo(1f, tween(holdDurationMs, easing = motion.linear))
                            haptics.performHapticFeedback(HapticFeedbackType.LongPress)
                            onHoldComplete()
                        }
                        tryAwaitRelease()
                        holdJob.cancel()
                        scope.launch { progress.animateTo(0f, tween(motion.quick)) }
                    },
                )
            },
        contentAlignment = Alignment.Center,
    ) {
        Box(
            modifier = Modifier
                .fillMaxHeight()
                .fillMaxWidth(fraction = progress.value.coerceIn(0f, 1f))
                .background(fillColor),
        )
        Text(
            text = text,
            fontFamily = CaprasimoFamily,
            color = contentColor,
        )
    }
}

File: iosApp/iosApp/Components/HoldToConfirmButton.swift
import SwiftUI
import Shared

/// The doc's "Hold to end" control: an outlined pill whose fill sweeps left-to-right over
/// `holdDurationMs` while pressed, firing `onHoldComplete` at 100%; releasing early cancels and
/// eases the fill back to zero. Ports `startHold`/`holdUp` from `HoldToConfirmButton.kt`.
struct HoldToConfirmButton: View {
    let text: String
    let holdDurationMs: Int32
    let onHoldComplete: () -> Void
    var minHeight: CGFloat = 56
    var borderColor: Color = Organic.colors.lockedOnSurfaceColor.organicAlpha(Organic.opacity.border)
    var fillColor: Color = Organic.colors.accentRamp.color800
    var contentColor: Color = Organic.colors.lockedOnSurfaceColor.organicAlpha(Organic.opacity.strong)

    @State private var progress: CGFloat = 0
    @State private var holdTask: Task<Void, Never>?

    var body: some View {
        GeometryReader { geometry in
            ZStack(alignment: .leading) {
                Rectangle()
                    .fill(fillColor)
                    .frame(width: geometry.size.width * progress)
                Text(text)
                    .font(.custom(OrganicFonts.display, fixedSize: 17))
                    .foregroundStyle(contentColor)
                    .frame(maxWidth: .infinity, alignment: .center)
            }
        }
        .frame(height: minHeight)
        .clipShape(Capsule())
        .overlay { Capsule().strokeBorder(borderColor, lineWidth: CGFloat(Organic.stroke.hairline)) }
        .contentShape(Capsule())
        .gesture(
            DragGesture(minimumDistance: 0)
                .onChanged { _ in if holdTask == nil { startHold() } }
                .onEnded { _ in cancelHold() }
        )
    }

    private func startHold() {
        progress = 0
        // `motion.linear` is the token's easing curve: the fill must track real elapsed time, so
        // a hold that looks 80% done really is 80% done.
        withAnimation(Organic.motion.linear.animation(durationMillis: holdDurationMs)) {
            progress = 1
        }
        holdTask = Task {
            try? await Task.sleep(nanoseconds: UInt64(holdDurationMs) * 1_000_000)
            guard !Task.isCancelled else { return }
            UIImpactFeedbackGenerator(style: .heavy).impactOccurred()
            onHoldComplete()
        }
    }

    private func cancelHold() {
        holdTask?.cancel()
        holdTask = nil
        withAnimation(.easeOut(duration: Organic.motion.quick.seconds)) { progress = 0 }
    }
}
